### ISSUE TITLE
Clarify all/any vs bitwise and/or

### ISSUE THREAD

**Issue by rebecca-palmer** [Theano/Theano#5495]
The [documentation](http://deeplearning.net/software/theano/library/tensor/elemwise.html) describes All/Any as bitwise and/or and CAReduce(and_)/CAReduce(or_) as all/any, and specifically states that these are equivalent.

However looking at the actual source, CAReduce(and_)/CAReduce(or_) are bitwise and/or, while All/Any are [bitwise and/or for int8 but logical and/or (i.e. what Numpy/Python call all/any) for all other types](https://github.com/Theano/theano/blob/84aa7becb8202ffce38d9fdca3cff2e47d39ae40/theano/tensor/elemwise.py#L1764).

test_CAReduce [tests All/Any against numpy all/any and and_/or_ against numpy bitwise_and/bitwise_or](https://github.com/Theano/Theano/blob/84aa7becb8202ffce38d9fdca3cff2e47d39ae40/theano/tensor/tests/test_elemwise.py#L384), but as [its integer test cases use only 0s and 1s,](https://github.com/Theano/Theano/blob/84aa7becb8202ffce38d9fdca3cff2e47d39ae40/theano/tensor/tests/test_elemwise.py#L360) it doesn't actually see this difference.

**Comment by lamblin** [comment]
You're right, there is something to be clarified or fixed, since we actually added a Boolean dtype quite recently (we used to use `int8` for that, which would be confusing).

I think the following would make sense:
- All/Any should cast their inputs as Booleans (or call `neq(..., 0)`), except for Booleans (but including int8, uint8)
- All/Any should return Booleans, not `int8`
- The documentation should be updated to state that Any/All are logical, and that a user could use CAReduce directly to have bitwise operations.
